javascript/Spread_operator.md

```diff
@@ -0,0 +1,198 @@
+```
+
+#### copying without the spread operator 
+
+```javaScript
+let arr = ['a','b','c']; 
+let arr2 = arr; 
+  
+console.log(arr2); // [ 'a', 'b', 'c' ] 
+```
+[From geeksforgeeks](https://www.geeksforgeeks.org/javascript-spread-operator/)
+
+### Spread in object literals
+
+It copies own enumerable properties from a provided object onto a new object.In object literals merging of objects is now possible using a shorter syntax than Object.assign().
+
+```javaScript
+const obj1 = { foo: 'bar', x: 42 };
+const obj2 = { foo: 'baz', y: 13 };
+
+const clonedObj = { ...obj1 };
+// Object { foo: "bar", x: 42 }
+
+const mergedObj = { ...obj1, ...obj2 };
+// Object { foo: "baz", x: 42, y: 13 }
+```
+
+## Array slice method
+
+The slice() method returns the selected elements in an array, as a new array object.The slice() method selects the elements starting at the given start argument, and ends at, but does not include, the given end argument.
+The **syntax** of slice method is `array.slice(start, end)`.
+
+### The Slice method takes 2 arguments.
+
+**1st Argument:** Specifies from where the selection should be started.
+
+For Example:
+
+```
+var arr1 = [1,5,8,9];
+arr1.slice(1); // [5,8,9]
+From the first index (5) it will return the elements.
+```
+
+**2nd Argument:** Specifies at which level the endpoint should be. If you didn’t put this in the parenthesis while calling the slice method, it will return the elements from the starting index to the end of the array.
+
+```
+var arr1 = [1,5,8,9];
+console.log(arr1.slice(1,3));
+//[ 5, 8 ]
+```
+
+If you put a negative number while calling, the selection will be selected from the end of the array.
+
+```
+var arr1 = [1,5,8,9];
+console.log(arr1.slice(-2));
+//[ 8, 9 ]
+```
+
+Note: Slice always returns the selected elements from the array.
+
+Slice won’t change the array. The array remains intact. See the below Example:
+
+```
+var arr1 = [1,5,8,9];
+arr1.slice(2);
+console.log(arr1);
+// [ 1, 5, 8, 9 ]
+```
+
+Even if you made some changes to the array it won’t affect it. It will return the original array as it is at the start.
+[From freecodecamp](https://www.freecodecamp.org/news/lets-explore-slice-splice-spread-syntax-in-javascript-e242a6f21e60/)
+
+
+## Independent Practice
+
+### Nested Object
+
+Rewrite the following code using spread syntax.
+
+```javaScript
+const arr1 = ['a','b', 'c'];
+const arr2 = ['a.foo', 'b.bar', 'c.baz'];
+
+let result = {};
+
+const insertAggs = (agg, result) => {
+
+  while (result.aggs) {
+    result = result.aggs
+  }
+
+  result.aggs = agg;
+}
+
+for (let i = 0; i < arr1.length; i++) {
+  const aggName = arr1[i];
+  const field = arr2[i];
+  const agg = {
+      [aggName]: {
+        terms: {
+          field
+        }
+      }
+    };
+
+  result.aggs ? insertAggs(agg, result) : result.aggs = agg;
+}
+}
+```
+[from freecodecamp](https://www.freecodecamp.org/forum/t/how-do-i-build-a-nested-object-in-javascript-dynamically/304543/3)
+
```

Comment:
```suggestion
## Challenge
```
Give them 2 word problems here, and tell them to write a function that uses spread, and one that uses slice to solve them. You are welcome to use problems you find as long as you cite the sources.

Reply:
I have already added some of the question in additional material. What can I add in challenges?

Reply:
1. tell them to write a function that uses spread as the solution
2.  tell them to write a function that uses slice as the solution

You don't need to make them up; but if you found it somewhere else, provide a link and say where it's from.

Reply:
Awesome!  Please add this to the Projected Time, too.